purchases/src/main/kotlin/com/revenuecat/purchases/storage/FileRepository.kt

```diff
@@ -0,0 +1,197 @@
+                        throw error
+                    }()
+
+                if (fileCacheManager.cachedContentExists(cachedUri)) {
+                    return@async cachedUri
+                }
+
+                val data = downloadFile(url = url)
+                saveCachedFile(cachedUri, data)
+                return@async cachedUri
+            }
+        }.await()
+    }
+
+    private suspend fun downloadFile(url: URL): InputStream {
+        return try {
+            withContext(Dispatchers.IO) {
+                verboseLog { "Downloading remote font from $url" }
+
+                val connection = urlConnectionFactory.createConnection(url.toString())
+
+                if (connection.responseCode != HttpURLConnection.HTTP_OK) {
+                    throw IOException("HTTP ${connection.responseCode} when downloading file at: $url")
+                }
+
+                return@withContext connection.inputStream
+            }
+        } catch (e: IOException) {
+            val message = "Failed to fetch file from remote source: $url. Error: ${e.localizedMessage}"
+            logHandler.e("FileRepository", message, e)
+            throw Error.FailedToFetchFileFromRemoteSource(message)
+        }
+    }
+
+    private fun saveCachedFile(uri: URI, data: InputStream) {
+        try {
+            fileCacheManager.saveData(data, uri)
+        } catch (e: IOException) {
+            val message = "Failed to save cached file: $uri. Error: ${e.localizedMessage}"
+            logHandler.e("FileRepository", message, e)
+            throw Error.FailedToSaveCachedFile(message)
+        }
+    }
+
+    /**
+     * File repository error cases.
+     */
+    sealed class Error(message: String) : IOException(message) {
+        /**
+         * Used when creating the folder on disk fails.
+         */
+        class FailedToCreateCacheDirectory(url: String) : Error("Failed to create cache directory for $url")
+
+        /**
+         * Used when saving the file on disk fails.
+         */
+        class FailedToSaveCachedFile(message: String) : Error(message)
+
+        /**
+         * Used when fetching the data fails.
+         */
+        class FailedToFetchFileFromRemoteSource(message: String) : Error(message)
+    }
+}
+
+private class FileCache(
+    private val context: Context,
+) : LargeItemCacheType {
+
+    private val md: MessageDigest by lazy {
+        MessageDigest.getInstance("MD5")
+    }
+
+    private val cacheDir: File by lazy {
+        val dir = File(context.cacheDir, "rc_files")
+        if (!dir.exists()) {
+            dir.mkdirs()
+        }
+        dir
+    }
+
+    override fun generateLocalFilesystemURI(remoteURL: URL): URI? {
+        val urlHash = md5Hex(remoteURL.toString().toByteArray())
+        val fileName = File(urlHash).name
+        if (fileName.isEmpty()) return null
+        return File(cacheDir, fileName).toURI()
+    }
+
+    override fun cachedContentExists(uri: URI): Boolean =
+        File(uri).exists()
```

Comment:
We discussed this but I'm still scared that we don't perform any verification that the download was performed successfully before saving the file... we could end up with a corrupted cache that would require uninstall/reinstall to come out from. I've seen this happen in fonts, and I fear it might happen more often on videos which are likely larger. Are we planning to add any sort of verification here so at least we don't end up stuck? @JZDesign @joshdholtz 

I know we talked about maybe calculate a checksum on the backend side that can be used to perform this verification... but not sure if there are better ways TBH.